### ISSUE TITLE
this library is not finished (are we there yet)

### ISSUE THREAD

**Issue by cdornan** [aelve/fmt#1]
This has been on my list of things to do for ages, but hadn't got round to it. I really like your solution &mdash; it is exactly the kind of thing that I was hoping to come up with.

(Also I think you have had some past involvement in the regex libraries &mdash; if you have any feedback or suggestions for [regex](http://hs.regex.uk) please give me a shout.)

**Comment by neongreen** [comment]
We're not there yet but I hope to finish it soon (e.g. today I'll spend the whole day on it).

**Comment by neongreen** [comment]
By the way, it's already on Hackage (http://hackage.haskell.org/package/fmt) and even kinda usable, though before releasing `0.1` I plan to introduce at least one breaking change – specifically, I want to change `%<foo>%` to `<%foo%>`. The former looks nicer, but somehow the latter is much easier to type, even though I've no idea why. (Is it the same for you or not?)

**Comment by cdornan** [comment]
Ah &mdash; I hadn't thought to check out Hackage &mdash; thanks!

I agree with you about the aesthetics, but there are good arguments for both styles &mdash; trust your own judgement and let the world suck it up.

As for breaking stuff, it is a real shame about the move away from the Hackage stability field as I think it can be really useful. I would warn folks that it is unstable and just break stuff at will (just as you are) until you are happy with it, slap a 1.0.0.0 (or 0.1) version on it and declare it stable.

**Comment by cdornan** [comment]
When I get the chance I will give it a spin and give you some feedback here.

**Comment by cdornan** [comment]
Just gone back to look at the Haddocks &mdash; love it! really easy for me to pick up the gist from a glance, and I like what I am seeing! Looks to me like Haskell could vault from the back of the class at this kind of thing to a stellar performer. Really looking forward to having a play now!

**Comment by neongreen** [comment]
> (Also I think you have had some past involvement in the regex libraries — if you have any feedback or suggestions for regex please give me a shout.)

I co-maintain regex-tdfa, but that's pretty much all. I skimmed `regex` docs, but got slightly scared because regexes are complicated :) I would have to take a look sometime later anyway, because I need to maintain https://guide.aelve.com/haskell/regular-expressions-jf3ju3u2, but I still kinda want to try my hand at making some library containing a dead-simple interface for regexes (in the future, maybe).

**Comment by ChShersh** [comment]
> specifically, I want to change `%<foo>%` to `<%foo%>`. The former looks nicer, but somehow the latter is much easier to type

I also noticed that. Somehow the second variant is easier for typing. But maybe you can consider some other syntax. Possible options from me:

:one:
```haskell 
"There are "%|n|%" million bicycles in "%|city|%"."
```
:two: (this option makes it consistent with `>|<` operator from README)
```haskell 
"There are "<|n|>" million bicycles in "<|city|>"."
```
:three: (unfortunately, `${` and `}$` are not possible names for operators :disappointed:)
```haskell 
"There are "$|n|$" million bicycles in "$|city|$"."
```
:four: 
```haskell 
"There are "#|n|#" million bicycles in "#|city|#"."
```
:five: (`%` and `^` very near on keyboard)
```haskell 
"There are "%^n^%" million bicycles in "%^city^%"."
```
I like :two: :four: :three: the most. But unfortunately :two: conflicts with `Data.Sequence` from `containers` :(


> but I still kinda want to try my hand at making some library containing a dead-simple interface for regexes

I really like Gabriel's [_regular expressions_ approach](https://github.com/Gabriel439/slides/blob/master/regex/regex.md). Unfortunately, not as library still, though all code is written. Maybe he is busy. But here is [the talk](https://begriffs.com/posts/2016-06-27-fast-haskell-regexes.html).

**Comment by neongreen** [comment]
Okay, I like `#|` more than `<%`, so I'll change it.

**Comment by neongreen** [comment]
We're finally there. (Released 0.1.)
